File: cap10/src/algoritmos/CompresionHuffman.java
package algoritmos;

import java.math.BigInteger;
import java.util.HashMap;
import java.util.Map;
import java.util.Map.Entry;
import java.util.PriorityQueue;

public class CompresionHuffman {
	
	private static final Character FIN = Character.MIN_VALUE;
	
	private Map<Character, Float> frecuencias = new HashMap<>();
	private PriorityQueue<Nodo> cola = new PriorityQueue<>();
	private Nodo arbol;
	private Map<Character, String> diccionarioCompresion = new HashMap<>();
	private Map<String, Character> diccionarioDescompresion = new HashMap<>();
	
	public CompresionHuffman(String textoFrecuencias) {
		frecuencias(textoFrecuencias);
		cola();
		arbol();
		diccionario(arbol);
	}

	private void frecuencias(String textoFrecuencias) {
		float total = textoFrecuencias.length() + 1;
		Map<Character, Integer> contador = new HashMap<>();
		for (char c : textoFrecuencias.toCharArray()) {
			int cont = 0;
			if (contador.containsKey(c)) {
				cont = contador.get(c);
			}
			contador.put(c, ++cont);
		}
		for (Entry<Character, Integer> e : contador.entrySet()) {
			frecuencias.put(e.getKey(), e.getValue()/total);
		}
		frecuencias.put(FIN, 1/total);
	}
	
	private void cola() {
		for (Entry<Character, Float> e : frecuencias.entrySet()) {
			Nodo n = new Nodo(e.getKey(), e.getValue());
			cola.add(n);
		}
	}
	
	private void arbol() {
		while (cola.size() > 1) {
			Nodo nodo1 = cola.poll();
			Nodo nodo2 = cola.poll();
			Nodo nuevoNodo = new Nodo(nodo1.getProbabilidad() 
					+ nodo2.getProbabilidad(), nodo1, nodo2);
			cola.add(nuevoNodo);
		}
		arbol = cola.poll();
	}
	
	private void diccionario(Nodo nodo) {
		if (nodo.getSimbolo() != null) {
			diccionarioCompresion.put(nodo.getSimbolo(), nodo.getCodigo());
			diccionarioDescompresion.put(nodo.getCodigo(), nodo.getSimbolo());
		} else {
			diccionario(nodo.getDerecho());
			diccionario(nodo.getIzquierdo());
		}
	}
	 
	public BigInteger comprimir(String texto) {
		String s = "1"; // que empiece por 1 para no perder ceros a la izquierda
		for (Character c : texto.toCharArray()) {
			s += diccionarioCompresion.get(c);
		}
		s += diccionarioCompresion.get(FIN);
		while (s.length() % 8 != 0) {
			s += "0";
		}
		return new BigInteger(s, 2);
	}
	
	public String descomprimir(BigInteger binario) {
		String comprimido = binario.toString(2);
		String s = "";
		boolean hecho = false;
		int i = 1; // ignoramos el primer uno
		while (i < comprimido.length() && !hecho) {
			String codigo = comprimido.charAt(i) + "";
			while (!diccionarioDescompresion.containsKey(codigo)) {
				codigo += comprimido.charAt(++i);
			}
			Character letra = diccionarioDescompresion.get(codigo);
			if (letra.equals(FIN)) {
				hecho = true;
			} else {
				s += letra;
				i++;
			}
		}
		
		return s;
	}
	
	public static void main(String[] args) {
		CompresionHuffman ch = new CompresionHuffman("ana anda y canta lalala.");
		BigInteger comprimido = ch.comprimir("ana anda y canta lalala.");
		System.out.println(comprimido);
		System.out.println(comprimido.toString(2));
		System.out.println(ch.descomprimir(comprimido));
		BigInteger comprimido2 = ch.comprimir("anda lana y cantala ya.");
		System.out.println(comprimido2);
		System.out.println(comprimido2.toString(2));
		System.out.println(ch.descomprimir(comprimido2));
	}
}

class Nodo implements Comparable<Nodo>{
	private Character simbolo;
	private Float prob;
	private Nodo izquierdo;
	private Nodo derecho;
	private Nodo padre;
	private boolean codigo;
	

	public Nodo(Character simbolo, Float prob) {
		this.simbolo = simbolo;
		this.prob = prob;
	}

	public Nodo(Float prob, Nodo izquierdo, Nodo derecho) {
		this.prob = prob;
		this.izquierdo = izquierdo;
		this.derecho = derecho;
		izquierdo.setPadre(this, true);
		derecho.setPadre(this, false);
	}
	
	public void setPadre(Nodo padre, boolean b) {
		this.padre = padre;
		this.codigo = b;
	}
	
	public Character getSimbolo() {
		return simbolo;
	}

	public String getCodigo() {
		return (padre != null ? padre.getCodigo() : "") + (codigo ? "1" : "0");
	}
	
	public Float getProbabilidad() {
		return prob;
	}
	
	public Nodo getIzquierdo() {
		return izquierdo;
	}

	public Nodo getDerecho() {
		return derecho;
	}

	@Override
	public int compareTo(Nodo o) {
		if (o == null) {
			throw new IllegalStateException(); 
		}
		return this.prob.compareTo(o.prob);
	}
	
	@Override
	public String toString() {
		return simbolo + ": " + prob + " - " + getCodigo();
	}
	
}

File: cap10/src/algoritmos/CompresionHuffman.py
import collections
import heapq

def probabilidades(contenido):
    total = len(contenido) + 1
    tablaHash = collections.Counter(contenido)
    res = {}
    for char, cont in tablaHash.items():
        res[char] = float(cont)/total
    res['end'] = 1.0/total
    return res

def arbol(probs):
    # Preparación de la cola, ordenada por prioridad y profundidad
    cola = []
    for simbolo, prob in probs.items():
        heapq.heappush(cola, (prob, 0, simbolo))

    # Construcción del árbol
    while len(cola) > 1:
        nodo1 = heapq.heappop(cola) # El símbolo menos probable
        nodo2 = heapq.heappop(cola) # El segundo menos probable

        # Creación del nuevo nodo, unión de los dos menos probables
        nuevoNodo = (
            nodo1[0] + nodo2[0], 
            max(nodo1[1], nodo2[1]) + 1, 
            [nodo1, nodo2])
        heapq.heappush(cola, nuevoNodo)
    return cola[0]
    
def diccionario(arbol):
    res = {}
    pilaBusqueda = []
    
    # El último elemento de la lista es el prefijo!
    pilaBusqueda.append(arbol + ("", )) 
    while len(pilaBusqueda) > 0:
        elem = pilaBusqueda.pop()
        if type(elem[2]) == list:
            # Nodo con hijos
            prefix = elem[-1]
            # El hijo izquierdo tiene "1" en el prefijo
            pilaBusqueda.append(elem[2][1] + (prefix+"1", ))
            # El hijo derecho tiene "0" en el prefijo
            pilaBusqueda.append(elem[2][0] + (prefix+"0", ))
            continue
        else:
            # Nodo hoja
            codigo = elem[-1]
            res[elem[2]] = codigo
        pass
    return res
    
def comprimir(dicc, contenido):
    res = "1"
    for letra in contenido:
        codigo = dicc[letra]
        res = res + codigo
    res = res + dicc['end']
    # Agregamos ceros para que la longitud del resultado sea un múltiplo de 8
    res = res + (len(res) % 8 * "0")
    return int(res, 2) # Convertimos el string en base 2 a entero
    
def descomprimir(dicc, comprimido):
    dic = dict(zip(dicc.values(), dicc.keys())) # giramos diccionario para decodificar
    res = []
    length = comprimido.bit_length() - 1
    if comprimido >> length != 1:
        raise Error("¡Fichero corrupto!")
    done = False
    while length > 0 and not done:
        shift = length - 1
        while True:
            num = comprimido >> shift
            bitnum = bin(num)[3:] # Quitamos '0b1' - el 1 inicial y el 0b de formato
            if bitnum not in dic:
                shift -= 1
                continue
            char = dic[bitnum]
            if char == 'end':
                done = True
                break
            res.append(char)
            comprimido = comprimido - ((num - 1) << shift)
            length = shift
    return ''.join(res)
    
def compresion(texto):
    probs = probabilidades(texto)
    arb = arbol(probs)
    dicc = diccionario(arb)
    res = comprimir(dicc, texto)
    rec = descomprimir(dicc, res)
    return res
    
    
def compresionVerbose(texto):
    probs = probabilidades(texto)
    print("probabilidades: ")
    print(probs)

    arb = arbol(probs)
    print("árbol: ")
    print(arb)

    dicc = diccionario(arb)
    print("diccionario:")
    print(dicc)
    
    res = comprimir(dicc, texto)

    rec = descomprimir(dicc, res) 
    print("descodificado: ")
    print(rec)
    return res
    
def bytesDe(i):
    return (i.bit_length() + 7) // 8    

texto0 = "CUANDO DIGO DIEGO, DIGO DIGO, CUANDO DIGO DIGO, DIGO DIEGO."
res0 = compresion(texto0)
print(res0)
print(str(len(texto0)) + " vs " + str(bytesDe(res0)))